src/lib/Web3Function.ts

```diff
@@ -88,10 +104,72 @@ export class Web3Function {
     }
   }
 
-  private async _run(ctxData: Web3FunctionContextData) {
-    if (!this._onRun)
+  private async _invokeOnRun(ctxData: Web3FunctionOnRunContextData) {
+    const context = this._context(ctxData);
+
+    if (ctxData.operation === "onRun" && !this._onRun)
       throw new Error("Web3Function.onRun function is not registered");
 
+    const result = ctxData.log
+      ? await (this._onRun as EventRunHandler)({
+          ...context,
+          log: ctxData.log,
+        })
+      : await (this._onRun as BaseRunHandler)(context);
+
+    return { result, ctxData };
+  }
+
+  private async _invokeOnFail(ctxData: Web3FunctionOnFailContextData) {
+    const context = this._context(ctxData);
+
+    if (ctxData.operation === "onFail" && !this._onFail)
+      throw new Error("Web3Function.onFail function is not registered");
+
+    if (ctxData.log) {
+      await (this._onFail as EventFailHandler)({
+        ...context,
+        reason: ctxData.onFailReason,
+        log: ctxData.log,
```

Comment:
I like the idea of passing back the `log` to the onFail & onSuccess handler but it might not be that practical from a backend perspective?
And I'm not too sure neither if the added value is actually worth it?